doc/developer/design/20220411_reclocking_implementation.md

```diff
@@ -0,0 +1,257 @@
+FROM KAFKA BROKER '...'
+EXPOSE RECLOCKING AS my_foo_clock
+```
+
+Materialize will make the contents of the `remap` collection for the source
+visible under the specified name via a relation whose structure depends on the
+type of the source:
+
+* File sources
+
+  | mz_timestamp | offset |
+  |--------------|--------|
+  | ...          | ...    |
+
+* Kafka sources
+
+  | mz_timestamp | partition | offset |
+  |--------------|-----------|--------|
+  | ...          | ...       | ...    |
+
+* Kinesis sources
+
+  | mz_timestamp | shard_id | sequence_number |
+  |--------------|----------|-----------------|
+  | ...          | ...      | ...             |
+
+* PostgreSQL sources
+
+  | mz_timestamp | lsn |
+  |--------------|-----|
+  | ...          | ... |
+
+* S3 sources
+
+  | mz_timestamp | key | offset |
+  |--------------|-----|--------|
+  | ...          | ... | ...    |
+
+
+If `EXPOSE RECLOCKING` is not specified, and the source type is reclockable,
+Materialize will generate a name of the form `<source>_reclocking`. If that name
+already exists, Materialize will try names of the form `<source>_reclocking1`,
+`<source>_reclocking2`, ... in sequence until it finds an available name. This
+matches the logic used for generatic the name of a primary index for a table.
+
+The reclocked relation will be a new catalog item of type `Reclocking`:
+
+```rust
+// catalog.rs
+/// An item representing a source reclocking relation.
+struct Reclocking {
+    /// The ID of the source to which this reclocking applies.
+    source_id: GlobalId,
+    /// The type of the reclocking relation.
+    ///
+    /// This must be derived the type of the `source_id` source. It is
+    /// duplicated here, rather than being derived from the source on the fly,
+    /// due to limitations in the design of the catalog API. (The type of the
+    /// item must be derivable from the item itself, i.e., without access to the
+    /// rest of catalog).
+    desc: RelationDesc,
+}
+```
+
+`Reclocking` items will necessarily have a `GlobalId` in order to be included in
+the catalog; however this `GlobalId` will not be meaningful to the storage
+layer. (TODO: should it be? Should we instead have `CreateSources` accept
+separate `GlobalId`s for the remap and reclocked collections, for the source
+types that support both? We'd need to ensure that `Insert, `AllowCompaction`,
+et al. are not supported for remap collections.)
+
+### System catalog changes
+
+A new system table `mz_reclockings` will be added with the following schema:
+
+| id | name | source_id |
+|----|------|-----------|
+| .. | ...  | ...       |
+
+This table relates the name of a reclocking collection to a source. There is no
+need to describe the database or schema of the reclocking collection; like
+indexes, they always belong to the same schema as the relation with which they
+are associated.
+
+The `mz_objects` view will include items from `mz_reclockings` with a type of
+`reclocking`.
+
+## Open questions
```

Comment:
Do we only ever need to support one remap collection?

Do we explicitly want, explicitly don't want, or want to punt on anything looking like byo support?

Reply:
I have thoughts here (that we want to support it) but I'm very happy to delay / dampen that in the interest of specific forward progress at the moment.

Reply:
Further spitballing: I think fundamentally the discussion with the user boils down to "to have your Kafka source in MZ it needs to be 'reclocked', whose input is a definite collection containing parts and offsets and stuff. We can create you a definite collection containing parts and offsets and stuff, though if you already have one that works too". I can totally imagine we implement it some other way, but if it turned out we implemented it this way, the separation of concerns would be super clear to me, at least.

Reply:
I'm a +1 on punting on BYO support, and we should explicitly state that in the doc (cc @benesch).

Reply:
I think it is probably useful to consider up front how STORAGE will re-ingest a Materialize-formatted sink. @frankmcsherry do you have thoughts on how reclocking looks in that situation, where we can trust the input to be definite?

Reply:
@frankmcsherry per our Slack conversation I added a section about how the implementation of this needs to be in two parts. PTAL!

Reply:
ITAL! @aljoscha's points about inputs that make the implementation hard for .. certain sources .. make sense! I'm not sure the right resolution there, other than a hard think about how excited we are about Kinesis sources. Do folks who use Kinesis sources have expectations of freshness / reclocking, given that presumably they too cannot see the tip of the stream without reading everything.